_posts/2019-12-05-testowanie-komponentow-z-inputami-i-outputami.md

```diff
@@ -0,0 +1,198 @@
+    resultText = fixture.debugElement.query(By.css('.result'));
+  }));
+
+  it('should show correct text when list is empty', () => {
+    component.savedValues = [];
+    fixture.detectChanges();
+    expect(resultText.childNodes[0].nativeNode.wholeText).toEqual('Brak zapisanych wartości');
+  });
+
+  it('should show elements of list with correct text', () => {
+    component.savedValues = [12, 3];
+    fixture.detectChanges();
+    expect(resultText.childNodes[0].nativeNode.wholeText).toEqual('Zapisane wartości to: 12,3');
+  });
+});
+```
+
+Zaczęliśmy tutaj, podobnie jak w poprzednim artykule, od stworzenia `TestBed` i dodatkowo wyciągnęliśmy za pomocą zapytania element html prezentujący tekst umieszczony z poziomu komponentu. Możemy się do niego odwołać na wiele sposobów, na przykład za pomocą klasy czy id.
+
+Dysponujemy tutaj dwoma testami o identycznej strukturze. Przekazujemy wartości w inpucie i sprawdzamy, czy element umieszczony w html został poprawnie zmieniony. Jeżeli inputów w komponencie mamy więcej, możemy je przetestować dokładnie w ten sam sposób.
+
+## Pisanie testów dotyczących @Output
+
+Następnie napiszmy test do komponentu pozwalającego wpisać liczbę, którą przekazuje do komponentu nadrzędnego.
+
+```typescript
+@Component({
+  selector: 'calculator-input-field',
+  template: `
+      <input (keyup)="onKey($event)">
+  `,
+})
+export class CalculatorInputFieldComponent {
+  @Output() fieldValue = new EventEmitter<number>();
+
+  onKey(event) {
+    this.fieldValue.emit(event.target.value);
+  }
+}
+```
+
+Komponent równie prosty jak poprzedni z takim wyjątkiem, że mamy tutaj do czynienia z outputem.
+
+```typescript
+describe('CalculatorInputFieldComponent', () => {
+  let component: CalculatorInputFieldComponent;
+  let fixture: ComponentFixture<CalculatorInputFieldComponent>;
+  let input: DebugElement;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [CalculatorInputFieldComponent]
+    }).compileComponents();
+
+    fixture = TestBed.createComponent(CalculatorInputFieldComponent);
+    component = fixture.componentInstance;
+    input = fixture.debugElement.query(By.css('input'));
+  }));
+
+  it('should emit correct value on input change', () => {
+    component.fieldValue.subscribe((value) => expect(value).toBe('7'));
+    input.nativeElement.value = 7;
+    document.querySelector('input').dispatchEvent(new KeyboardEvent('keyup'));
+  });
+});
+```
+
+Czyli standardowo, zaczynamy od przygotowania naszego modułu testowego i uchwycenia elementu, który chcemy przetestować. Warto tutaj zaznaczyć, że istnieją inne sposoby na zainicjowanie komponentu niż `TestBed`, wszystko zależy od tego co, i w jaki sposób, będziemy testować. Na przykład kiedy nie potrzebujemy testować elementów z DOM możemy zostać przy zwykłym inicjalizowaniu komponentów. Więcej szczegółów można znaleźć w [dokumentacji od Angulara](https://angular.io/guide/testing#component-test-basics).
```

Comment:
Dodatkowo dokumentacja jest raczej do czegoś, niż od czegoś, a w ogóle najlepiej to skróciłbym do `dokumentacji Angulara` :)